### DISCUSSION TITLE
kart reset --soft

### DISCUSSION THREAD

**Discussion by SrNetoChan** [koordinates/kart#626] [Ideas]
Following the idea that who knows git, knows kart. I miss one particular feature.

`git reset <commit>` uses by default the --soft tag, which reverts the HEAD to the desired commit but keeps the changes. which can then be discarded or not.

kart reset uses the --hard approach, which means that it will mode the HEAD losing all the changes made.

I couldn't find a way to do the soft approach... even using other commands. I would be nice if there was a --hard option to clean the changes and that the reset would default to --soft.

**Comment by rcoup** [comment]
Hi @SrNetoChan,

We've been reluctant to automatically throw every feature of Git into Kart: Git is complex and it's UX is too. So it's not that we're _against_ it, just wary. Similarly, we may end up with an index in future but we've avoided it so far.

Can you explain some example use cases where you've missed `reset --soft`?

**Reply by SrNetoChan** [reply]
I understand your concern in keeping it simple. Moreover, some functionality doesn't make sense with spatial data as it does with plain text

The use case is the following:

I do some changes in layer, and I commit them to the repository. Later, I realize I made a few mistakes in some features and while I want to revert the changes on those particular features, I don't want to loose all the changes I did on the rest.

If there was a `kart reset --soft`, I could undo the commit, keeping its changes in the working copy (staging in git). Then, I could revisit the features, discard the changes in the problematic features and then commit the rest of the changes again.

Maybe there's already a way of doing this, but I couldn't catch it.

Thanks!

**Comment by olsen232** [comment]
Sorry I haven't been keeping up with discussions here... the workaround at present would be to create a patch. 
1. You are at commit GOOD
2. You create commit PROBLEM-CHANGE that has GOOD as a parent.
3. You use kart reset to go back to GOOD state: `kart reset HEAD^`
4. You use `kart create-patch PROBLEM-CHANGE > mypatch` to create a patch that applies the same changes.
5. Use `kart apply --no-commit mypatch` to reapply but not commit the changes.
